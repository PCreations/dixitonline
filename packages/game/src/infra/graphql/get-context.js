export const makeGetContext = ({ dispatchDomainEvents, authorizationService } = {}) => async context => ({
  dispatchDomainEvents,
  currentUser: await authorizationService.getCurrentUser(context),
});
