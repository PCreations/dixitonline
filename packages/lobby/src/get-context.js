export const makeGetContext = ({ dispatchDomainEvents } = {}) => () => ({
  dispatchDomainEvents,
});
