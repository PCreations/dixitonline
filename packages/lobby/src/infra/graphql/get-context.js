export const makeGetContext = ({ dispatchDomainEvents, authorizationService } = {}) => async ({ req }) => ({
  dispatchDomainEvents,
  currentUser: await authorizationService.getCurrentUser(req),
});
