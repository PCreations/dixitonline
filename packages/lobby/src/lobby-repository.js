import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
          };
        },
      };
    },
  };
};
