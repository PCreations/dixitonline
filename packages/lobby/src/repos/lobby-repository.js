import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
    getAllGames() {
      return lobbyGames.get().then(snapshot => {
        const games = [];
        snapshot.forEach(doc => games.push(doc.data()));
        return games;
      });
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
            set(game) {
              gamesData[gameId] = game;
            },
          };
        },
        async get() {
          const docs = Object.values(gamesData).map(game => ({
            data() {
              return game;
            },
          }));
          return {
            forEach: docs.forEach.bind(docs),
          };
        },
      };
    },
  };
};
