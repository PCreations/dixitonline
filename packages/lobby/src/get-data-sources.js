import { makeLobbyRepository } from './lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
