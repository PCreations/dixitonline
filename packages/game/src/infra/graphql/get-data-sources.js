import { makeGameRepository } from '../../repos/game-repository';

export const makeGetDataSources = ({ gameRepository = makeGameRepository() } = {}) => () => ({
  gameRepository,
});
