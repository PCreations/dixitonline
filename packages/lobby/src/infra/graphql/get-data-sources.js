import { makeLobbyRepository } from '../../repos/lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
