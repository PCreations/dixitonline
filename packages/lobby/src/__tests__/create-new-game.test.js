import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import { makeServer } from './test-server';
import { makeNullLobbyRepository } from '../lobby-repository';
import { makeGetDataSources } from '../get-data-sources';
import { makeGetContext } from '../get-context';
import { newGameCreatedEvent } from '../domain-events';

describe('create new game', () => {
  test('it creates a new game', async () => {
    // arrange
    const dispatchDomainEvents = jest.fn();
    const server = makeServer({
      getDataSources: makeGetDataSources({
        lobbyRepository: makeNullLobbyRepository({ nextGameId: 'g1' }),
      }),
      getContext: makeGetContext({ dispatchDomainEvents }),
    });
    const CREATE_GAME = gql`
      mutation CreateGame {
        createGame {
          game {
            id
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);

    // assert
    const response = await mutate({ mutation: CREATE_GAME });
    expect(response).toMatchSnapshot();
    expect(dispatchDomainEvents).toHaveBeenCalledWith([newGameCreatedEvent({ gameId: 'g1' })]);
  });
});
