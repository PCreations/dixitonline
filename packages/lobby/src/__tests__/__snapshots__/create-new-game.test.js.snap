// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create new game it creates a new game 1`] = `
Object {
  "data": Object {
    "createGame": Object {
      "game": Object {
        "id": "g1",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
